frontend/src/plugins/impl/vega/VegaPlugin.tsx

```diff
@@ -0,0 +1,155 @@
+      .transform((spec) => spec as unknown as VegaLiteSpec),
+    chartSelection: z
+      .union([z.boolean(), z.literal("point"), z.literal("interval")])
+      .default(true),
+    fieldSelection: z.union([z.boolean(), z.array(z.string())]).default(true),
+  });
+
+  render(props: IPluginProps<T, Data>): JSX.Element {
+    return (
+      <VegaComponent
+        value={props.value}
+        setValue={props.setValue}
+        {...props.data}
+      />
+    );
+  }
+}
+
+interface VegaComponentProps<T> extends Data {
+  value: T;
+  setValue: (value: T) => void;
+}
+
+export const VegaComponent = ({
+  value,
+  setValue,
+  chartSelection,
+  fieldSelection,
+  spec,
+}: VegaComponentProps<T>): JSX.Element => {
+  const vegaView = useRef<View>();
+  const [error, setError] = useState<Error>();
+
+  // Debug
+  useDebugMounting("VegaComponent");
+  usePropsDidChange("VegaComponent", {
+    value,
+    setValue,
+    chartSelection,
+    fieldSelection,
+    spec,
+  });
+
+  // Aggressively memoize the spec, so Vega doesn't re-render/re-mount the component
+  const selectableSpec = useMemo(() => {
+    return makeSelectable(spec, { chartSelection, fieldSelection });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [useDeepCompareMemoize(spec), chartSelection, fieldSelection]);
+  const names = useMemo(
+    () => getSelectionParamNames(selectableSpec),
+    [selectableSpec]
+  );
+
+  // Update the value when the selection changes
+  // Keep the old value from other signals
+  const handleUpdateValue = useEvent((newValue: T) => {
+    setValue({ ...value, ...newValue });
+  });
+
+  const signalListeners = useMemo(
+    () =>
+      names.reduce((acc, name) => {
+        // Debounce each signal listener, otherwise we may create expensive requests
+        acc[name] = debounce((signalName, signalValue) => {
+          Logger.debug("[Vega signal]", signalName, signalValue);
+          handleUpdateValue({
+            [signalName]: signalValue,
+          });
+        }, 100);
+        return acc;
+      }, {} as SignalListeners),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [useDeepCompareMemoize(names), setValue]
+  );
+
+  const handleError = useEvent((error) => {
+    Logger.error(error);
+    Logger.debug(selectableSpec);
+    setError(error);
+  });
+
+  const handleNewView = useEvent((view) => {
+    Logger.debug("[Vega view] created", view);
+    vegaView.current = view;
+    setError(undefined);
+  });
+
+  return (
+    <>
+      {error && (
+        <Alert variant="destructive">
+          <AlertTitle>{error.message}</AlertTitle>
+          <div className="text-md">{error.stack}</div>
+        </Alert>
+      )}
+      <VegaLite
+        spec={selectableSpec}
+        actions={actions}
+        signalListeners={signalListeners}
+        onError={handleError}
+        onNewView={handleNewView}
+      />
+    </>
+  );
+};
+
+const actions = {
+  source: false,
+  compiled: false,
+};
```

Comment:
bunch of debug stuff in here - its all ignored in production, so planned to leave it in